docs/codeql/codeql-language-guides/analyzing-data-flow-in-cpp-new.rst

```diff
@@ -0,0 +1,431 @@
+.. _analyzing-data-flow-in-cpp-new:
+
+.. pull-quote:: Note
+
+   The data flow library described here is available from CodeQL 2.13.0 onwards. See :ref:`here <analyzing-data-flow-in-cpp>` for the library available in earlier versions.
+
+Analyzing data flow in C and C++
+================================
+
+You can use data flow analysis to track the flow of potentially malicious or insecure data that can cause vulnerabilities in your codebase.
+
+About data flow
+---------------
+
+Data flow analysis computes the possible values that a variable can hold at various points in a program, determining how those values propagate through the program, and where they are used. In CodeQL, you can model both local data flow and global data flow. For a more general introduction to modeling data flow, see ":ref:`About data flow analysis <about-data-flow-analysis>`."
+
+Local data flow
+---------------
+
+Local data flow is data flow within a single function. Local data flow is usually easier, faster, and more precise than global data flow, and is sufficient for many queries.
+
+Using local data flow
+~~~~~~~~~~~~~~~~~~~~~
+
+The local data flow library is in the module ``DataFlow``, which defines the class ``Node`` denoting any element that data can flow through. ``Node``\ s are divided into expression nodes (``ExprNode``) and parameter nodes (``ParameterNode``). It is possible to map between data flow nodes and expressions/parameters using the member predicates ``asExpr``, ``asIndirectExpr``, and ``asParameter``:
```

Comment:
This probably still needs to be fixed. `ExprNode` and `ParameterNode` only make partial sense. It was not clear to me what I should replace this with.

Reply:
Partial sense in that I wonder if the the indirect nodes can be categorised as `ExprNode`s and `ParameterNode`s, or are they somehow in separate categories which don't have a clear representation from the user's perspective?

Reply:
I've added some more text here.

Reply:
Looks good to me 👍